test/gtest/uct/test_mem.cc

```diff
@@ -106,25 +108,42 @@ UCS_TEST_P(test_mem, md_alloc) {
         status = uct_md_query(md, &md_attr);
         ASSERT_UCS_OK(status);
 
-        for (nonblock = 0; nonblock <= 1; ++nonblock) {
-            int flags = nonblock ? UCT_MD_MEM_FLAG_NONBLOCK : 0;
-
-            flags         |= UCT_MD_MEM_ACCESS_ALL;
-            params.flags   = flags;
-            params.mds.mds = &md;
-
-            status = uct_mem_alloc(length, methods, 3, &params, &mem);
-            ASSERT_UCS_OK(status);
-
-            if (md_attr.cap.flags & UCT_MD_FLAG_ALLOC) {
-                EXPECT_EQ(UCT_ALLOC_METHOD_MD, mem.method);
-            } else {
-                EXPECT_NE(UCT_ALLOC_METHOD_MD, mem.method);
+        for (mem_type = UCS_MEMORY_TYPE_HOST; mem_type < UCS_MEMORY_TYPE_LAST; mem_type++) {
+            params.mem_type = (ucs_memory_type_t)mem_type;
+            if (md_attr.cap.alloc_mem_types & UCS_BIT(mem_type)) {
+                for (nonblock = 0; nonblock <= 1; ++nonblock) {
+                    if ((nonblock == 1) && (mem_type != UCS_MEMORY_TYPE_HOST)) {
+                        continue;
+                    }
+                    if ((mem_type == UCS_MEMORY_TYPE_CUDA)
+                        || (mem_type == UCS_MEMORY_TYPE_CUDA_MANAGED)) {
+                        if (!mem_type_ctx_created) {
+                            address = mem_buffer::allocate(1, (ucs_memory_type_t)mem_type);
+                            mem_buffer::release(address, (ucs_memory_type_t)mem_type);
```

Comment:
why needed allocate+release?

Reply:
This is has the effect of initializing cuda context which is needed for uct_mem_alloc on non-host mem_type to work.

Reply:
can we use cudaSetDevice()? also can we move this to init function?

Reply:
Using cudaSetDevice is possible but that would bring cuda specific code into the test.

Also, I had tried to move this to init but I had seen that I was running into test failures because initialization didn't occur in time.

Reply:
isn't Cuda already initialized as part of UCT?

Reply:
Looks like today `cudaSetDevice`/`cuInit` is called from gtest/ucm/cuda_hooks.cc.

An alternative is to move allocate/release into gtest/common/mem_buffer.cc under say `mem_buffer::init_mem_type_ctx(mem_type)` and call that where necessary. This can be called at the common initialization function for UCT tests if there exists one.  Can you point to that function? There are a lot of init functions under gtest/uct.

Reply:
cuda init should be part of cuda_copy MD initialization . if we have UCT md object which claims to allocate cuda memory, it should have initialized everything needed for such allocation.

Reply:
> isn't Cuda already initialized as part of UCT?

Cuda initialization is user's responsibility and cuda_copy UCT reuses an existing context. 

> cuda init should be part of cuda_copy MD initialization

isn't initialization happening just before the test here - https://github.com/openucx/ucx/blob/a6c475b8604c640621dd37699610d917377c6de5/test/gtest/uct/test_mem.cc#L104

>  if we have UCT md object which claims to allocate cuda memory, it should have initialized everything needed for such allocation.

Unlike other mds which allocate, cuda_copy_md requires a context be present for allocation at the right location (device). Technically, we can allocate even if we don't have a context in cuda_copy UCT but that would be on default device 0 and this may be undesirable for all processes other than the process bound to device 0. So we fail today if there is no active context https://github.com/openucx/ucx/pull/5737/files#diff-f990528a14758f5a40bc106bd5fb5888513d667e4fc0b80256fe09dee83015c0R141.  

After discussing with @bureddy  a bit, if introducing `mem_buffer::init_mem_type_ctx(mem_type)` seems reasonable, I can add that to common and use that instead of explicitly allocating and releasing as I'm doing in the test.

Reply:
@yosefe any further thoughts on which direction we should take?

Reply:
> Cuda initialization is user's responsibility and cuda_copy UCT reuses an existing context.

So if user initializes cuda context only after MPI_Init, it's possible uct_md will not be able to allocate bounce buffers during MPI_Init ?

Reply:
Correct. The goal is to allocate lazily at the point when user requests a transfer from device buffer anyway. So having a context present at MPI_Init time is not a requirement and also immaterial. The first ucs_mpool_get will trigger allocation call and by this time, there should be a device context present.

Reply:
seems a very indirect assumption about behavior which may change..

> Technically, we can allocate even if we don't have a context in cuda_copy UCT but that would be on default device 0
Any reason not to use device 0 in gtest?

Reply:
We can do that if we only plan to test device 0 in gtest. Is that a safe assumption? Don't we have multiple devices on nodes in the testing infrastructure?